<img align="right" style="padding-left:50px;" src="figures_wk4/data_cleaning.png" width=350><br>
### User Bias in Data Cleaning
For your homework assignment this week, we will explore how our treatment of our data can impact the quality of our results.

**Dataset:**
The data is a Salary Survey from AskAManager.org. It’s US-centric-ish but does allow for a range of country inputs.

A list of the corresponding survey questions can be found [here](https://www.askamanager.org/2021/04/how-much-money-do-you-make-4.html).

 

In [18]:
import pandas as pd

In [19]:
df= pd.read_csv('survey_data.csv')

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28108 entries, 0 to 28107
Data columns (total 18 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   timestamp  28108 non-null  object 
 1   q1         28108 non-null  object 
 2   q2         28033 non-null  object 
 3   q3         28107 non-null  object 
 4   q4         7273 non-null   object 
 5   q5         28108 non-null  object 
 6   q6         20793 non-null  float64
 7   q7         28108 non-null  object 
 8   q8         211 non-null    object 
 9   q9         3047 non-null   object 
 10  q10        28108 non-null  object 
 11  q11        23074 non-null  object 
 12  q12        28026 non-null  object 
 13  q13        28108 non-null  object 
 14  q14        28108 non-null  object 
 15  q15        27885 non-null  object 
 16  q16        27937 non-null  object 
 17  q17        27931 non-null  object 
dtypes: float64(1), object(17)
memory usage: 3.9+ MB


In [21]:
df.head()

timestamp     q1                             q2  \
0  4/27/2021 11:02:10  25-34   Education (Higher Education)   
1  4/27/2021 11:02:22  25-34              Computing or Tech   
2  4/27/2021 11:02:38  25-34  Accounting, Banking & Finance   
3  4/27/2021 11:02:41  25-34                     Nonprofits   
4  4/27/2021 11:02:42  25-34  Accounting, Banking & Finance   

                                         q3   q4      q5      q6   q7   q8  \
0        Research and Instruction Librarian  NaN  55,000     0.0  USD  NaN   
1  Change & Internal Communications Manager  NaN  54,600  4000.0  GBP  NaN   
2                      Marketing Specialist  NaN  34,000     NaN  USD  NaN   
3                           Program Manager  NaN  62,000  3000.0  USD  NaN   
4                        Accounting Manager  NaN  60,000  7000.0  USD  NaN   

    q9             q10             q11          q12           q13  \
0  NaN   United States   Massachusetts       Boston     5-7 years   
1  NaN  United Kingdom             NaN    Cambridge  8 - 10 years   
2  NaN              US       Tennessee  Chattanooga   2 - 4 years   
3  NaN             USA       Wisconsin    Milwaukee  8 - 10 years   
4  NaN              US  South Carolina   Greenville  8 - 10 years   

           q14              q15         q16    q17  
0    5-7 years  Master's degree       Woman  White  
1    5-7 years   College degree  Non-binary  White  
2  2 - 4 years   College degree       Woman  White  
3    5-7 years   College degree       Woman  White  
4    5-7 years   College degree       Woman  White

### Assignment
Your goal for this assignment is to observe how your data treatment during the cleaning process can skew or bias the dataset.

Before diving right in, stop and read through the questions associated with the dataset. As you can see, they are either free-form text entries or categorical selections. Knowing this, perform some exploratory data analysis (EDA) to investigate the "state" of the dataset.

[Add as many code cell below here as needs]


### I'm going to get a bit more organized with my inspection of the data by listing out the questions as q1, q2, etc. make it challenging to see what I'm looking at.

##### 1. How old are you?
##### 2. What industry do you work in?
##### 3. Job title
##### 4. If your job title needs additional context, please clarify here:
##### 5. What is your annual salary?
##### 6. How much additional monetary compensation do you get, if any?
##### 7. Please indicate currency
##### 8. If "other," please indicate the currency here:
##### 9. If your income needs additional context, please provide here:
##### 10. What country do you work in?
##### 11. If you're in the U.S., what state do you work in?
##### 12. What city do you work in?
##### 13. How many years of professional work experience do you have overall?
##### 14. How many years of professional experience do you have in your field?
##### 15. What is your highest level of education completed?
##### 16. What is your gender?
##### 17. What is your race?

In [22]:
df.describe

<bound method NDFrame.describe of                  timestamp     q1                             q2  \
0       4/27/2021 11:02:10  25-34   Education (Higher Education)   
1       4/27/2021 11:02:22  25-34              Computing or Tech   
2       4/27/2021 11:02:38  25-34  Accounting, Banking & Finance   
3       4/27/2021 11:02:41  25-34                     Nonprofits   
4       4/27/2021 11:02:42  25-34  Accounting, Banking & Finance   
...                    ...    ...                            ...   
28103  12/29/2024 23:50:26  25-34                  Entertainment   
28104    1/3/2025 16:03:10  18-24   Engineering or Manufacturing   
28105   1/16/2025 10:00:47  35-44                  Entertainment   
28106   1/23/2025 19:38:04  35-44                     Nonprofits   
28107   1/27/2025 21:02:19  45-54              Computing or Tech   

                                                   q3   q4       q5       q6  \
0                  Research and Instruction Librarian  NaN   55,000  

In [23]:
df['q5'].describe()

count      28108
unique      4319
top       60,000
freq         430
Name: q5, dtype: object

In [24]:
df.isnull()

timestamp     q1     q2     q3    q4     q5     q6     q7     q8    q9  \
0          False  False  False  False  True  False  False  False   True  True   
1          False  False  False  False  True  False  False  False   True  True   
2          False  False  False  False  True  False   True  False   True  True   
3          False  False  False  False  True  False  False  False   True  True   
4          False  False  False  False  True  False  False  False   True  True   
...          ...    ...    ...    ...   ...    ...    ...    ...    ...   ...   
28103      False  False  False  False  True  False   True  False   True  True   
28104      False  False  False  False  True  False  False  False   True  True   
28105      False  False  False  False  True  False   True  False  False  True   
28106      False  False  False  False  True  False   True  False   True  True   
28107      False  False  False  False  True  False  False  False   True  True   

         q10    q11    q12    q13    q14    q15    q16    q17  
0      False  False  False  False  False  False  False  False  
1      False   True  False  False  False  False  False  False  
2      False  False  False  False  False  False  False  False  
3      False  False  False  False  False  False  False  False  
4      False  False  False  False  False  False  False  False  
...      ...    ...    ...    ...    ...    ...    ...    ...  
28103  False  False  False  False  False  False  False  False  
28104  False  False  False  False  False  False  False  False  
28105  False   True  False  False  False  False  False  False  
28106  False  False  False  False  False  False  False  False  
28107  False  False  False  False  False  False  False  False  

[28108 rows x 18 columns]

In [25]:
df['q8'].isnull().sum()

np.int64(27897)

In [26]:
df['q9'].isnull().sum()

np.int64(25061)

In [27]:
df['q8'].value_counts()

q8
INR                                       11
SGD                                       11
USD                                       11
NOK                                       10
DKK                                        8
                                          ..
US Dollar                                  1
AUD & NZD are not the same currency...     1
55,000                                     1
ILS/NIS                                    1
TZS                                        1
Name: count, Length: 124, dtype: int64

In [28]:
df['q7'].value_counts()

q7
USD        23410
CAD         1675
GBP         1592
EUR          646
AUD/NZD      504
Other        164
CHF           37
SEK           37
JPY           23
ZAR           16
HKD            4
Name: count, dtype: int64

In [29]:
df['q10'].value_counts()

q10
United States      9004
USA                7946
US                 2612
Canada             1572
United States       668
                   ... 
IS                    1
United Kingdomk       1
 New Zealand          1
Cuba                  1
Česká republika       1
Name: count, Length: 382, dtype: int64

In [30]:
df['q4'].value_counts().sum()

np.int64(7273)

In [31]:
df.head(30)

timestamp     q1                                    q2  \
0   4/27/2021 11:02:10  25-34          Education (Higher Education)   
1   4/27/2021 11:02:22  25-34                     Computing or Tech   
2   4/27/2021 11:02:38  25-34         Accounting, Banking & Finance   
3   4/27/2021 11:02:41  25-34                            Nonprofits   
4   4/27/2021 11:02:42  25-34         Accounting, Banking & Finance   
5   4/27/2021 11:02:46  25-34          Education (Higher Education)   
6   4/27/2021 11:02:51  25-34                            Publishing   
7   4/27/2021 11:03:00  25-34         Education (Primary/Secondary)   
8   4/27/2021 11:03:01  45-54                     Computing or Tech   
9   4/27/2021 11:03:02  35-44         Accounting, Banking & Finance   
10  4/27/2021 11:03:03  25-34                            Nonprofits   
11  4/27/2021 11:03:07  35-44          Education (Higher Education)   
12  4/27/2021 11:03:09  35-44         Accounting, Banking & Finance   
13  4/27/2021 11:03:10  25-34                                   Law   
14  4/27/2021 11:03:11  18-24                           Health care   
15  4/27/2021 11:03:11  35-44        Utilities & Telecommunications   
16  4/27/2021 11:03:13  35-44                Business or Consulting   
17  4/27/2021 11:03:19  45-54                          Art & Design   
18  4/27/2021 11:03:21  35-44                Business or Consulting   
19  4/27/2021 11:03:23  35-44          Education (Higher Education)   
20  4/27/2021 11:03:24  25-34                           Health care   
21  4/27/2021 11:03:24  35-44                            Nonprofits   
22  4/27/2021 11:03:27  35-44                            Nonprofits   
23  4/27/2021 11:03:29  35-44  Government and Public Administration   
24  4/27/2021 11:03:31  25-34                        Public Library   
25  4/27/2021 11:03:34  35-44          Education (Higher Education)   
26  4/27/2021 11:03:38  25-34                            Nonprofits   
27  4/27/2021 11:03:40  35-44                                   Law   
28  4/27/2021 11:03:41  35-44  Government and Public Administration   
29  4/27/2021 11:03:43  25-34          Engineering or Manufacturing   

                                                   q3  \
0                  Research and Instruction Librarian   
1            Change & Internal Communications Manager   
2                                Marketing Specialist   
3                                     Program Manager   
4                                  Accounting Manager   
5                      Scholarly Publishing Librarian   
6                                Publishing Assistant   
7                                           Librarian   
8                                     Systems Analyst   
9                                   Senior Accountant   
10                                     Office Manager   
11  Deputy Title IX Coordinator/ Assistant Directo...   
12                    Manager of Information Services   
13                           Legal Aid Staff Attorney   
14                           Patient care coordinator   
15                  Quality And Compliance Specialist   
16                                Executive Assistant   
17                                   graphic designer   
18                                     Senior Manager   
19            Assistant Director of Academic Advising   
20                            Data Programmer Analyst   
21            Program Coordinator & Assistant Editor    
22                                      Event Planner   
23                                         Researcher   
24                                    Teen Librarian    
25                          Communications Specialist   
26                                   Program Director   
27                         Bookkeeper/Billing Manager   
28                                          Economist   
29                                  Research Engineer   

                                           

In [32]:
df.tail(30)

timestamp          q1                                    q2  \
28078   9/23/2024 17:40:20       18-24              Leisure, Sport & Tourism   
28079   9/23/2024 22:53:33       35-44                           Health care   
28080   9/25/2024 19:54:18       45-54         Accounting, Banking & Finance   
28081   9/25/2024 21:34:43       45-54         Education (Primary/Secondary)   
28082   9/30/2024 10:52:30       55-64  Government and Public Administration   
28083   10/4/2024 13:05:24       18-24                     Computing or Tech   
28084    10/7/2024 5:17:56  65 or over                       Social networks   
28085  10/28/2024 17:33:38       35-44                             Insurance   
28086  10/30/2024 13:44:49       35-44                           Health care   
28087   11/7/2024 15:24:30       18-24                     Computing or Tech   
28088   11/7/2024 15:36:52       18-24                     Computing or Tech   
28089   11/11/2024 7:28:08       18-24           Marketing, Advertising & PR   
28090  11/12/2024 13:12:37       18-24           Marketing, Advertising & PR   
28091  11/16/2024 13:53:53       25-34          Engineering or Manufacturing   
28092  11/16/2024 13:55:56       25-34          Engineering or Manufacturing   
28093   11/18/2024 7:24:09       45-54                     Computing or Tech   
28094  11/18/2024 11:08:58       18-24                     Computing or Tech   
28095  11/22/2024 15:35:09       25-34                  Hospitality & Events   
28096   11/24/2024 8:08:37       25-34                Business or Consulting   
28097   12/5/2024 17:12:41       25-34         Education (Primary/Secondary)   
28098    12/6/2024 5:45:00       18-24                                   NaN   
28099   12/6/2024 19:52:27       35-44  Government and Public Administration   
28100  12/11/2024 14:53:19       25-34                           Health care   
28101  12/16/2024 19:10:58       18-24          Engineering or Manufacturing   
28102   12/20/2024 6:16:27       55-64  Government and Public Administration   
28103  12/29/2024 23:50:26       25-34                         Entertainment   
28104    1/3/2025 16:03:10       18-24          Engineering or Manufacturing   
28105   1/16/2025 10:00:47       35-44                         Entertainment   
28106   1/23/2025 19:38:04       35-44                            Nonprofits   
28107   1/27/2025 21:02:19       45-54                     Computing or Tech   

                                                   q3  \
28078                                        dockhand   
28079                                  Pharmacy Buyer   
28080                            Product Manager Lead   
28081                               Curriculum Writer   
28082                                Clerical Officer   
28083                      Software Engineering Co-Op   
28084                                 Content creator   
28085                                   Claim Manager   
28086                                        vcyuiugf   
28087                           Direct Support worker   
28088                                Business Analyst   
28089                      email marketing specialist   
28090                      Email Marketing Specialist   
28091                      Reconditioning associate 2   
28092                      Reconditioning Asscoiate 2   
28093                                  CYBER SECURITY   
28094                       Cyber Security Management   
28095                                 Revenue Manager   
28096                                  BI Consultant    
28097                          Elementary Art Teacher   
28098                                   data engineer   
28099                               Station Commander   
28100                                      Pharmacist   
28101      Clinical Specialist - Biomedical Engineer    
28102                            HMRC Engagement Lead   
28103                                   Junior Editor 

**Question:** How would you describe the "state" of this dataset? Be specific and detailed in your answer. (Think paragraphs rather than sentences).

[Add you answer to this markdown cell]

I would describe the state of this dataset as messy. The first problem I see is that there are a very large number of null values across several different columns. When looking at the survey we can see exactly how that happened as not all of features are required to be filled out to complete the survey. I can understand why gender and race would be optional fields but not sure why State is not a required field. There are also several questions that are conditional based on the response on the previous question, and are still not required fields. There are so few responses to the conditional questions that I'm not sure we can find any real value or include them as part of our analysis.

Another major problem that we have is that salaries are not all listed in one currency. We will have to engineer at least one new feature to create any kind of analysis based on salary, which I think is the objective of this kind of data. Another problem is that there is an option to free form type your currency. Because of that, we have all sorts of different descriptions for currency types, including many overlapping with other currencies. Some people even typed out "US Dollar" instead of selecting "USD" from the drop menu.

Going back to the conditional questions, they have created a massive source of null values. If the first question is left unanswered that creates a null value, and if the first question is answered, it creates a null value for the conditional question field.

#### The Plan

Now, it is time to plan how you will clean up the dataset. You **are not** allowed to use any machine learning technique to clean the data. (No SMOTE! No machine learning! Or anything like that!)

**Question:** Based on your EDA above, detail how you would clean up this dataset. 
Things to consider: (This is not an exhaustive list)
- Are there columns that can't be effectively cleaned? If so, why?
- Are there columns that genuinely won't have a data value?
- Does it make sense to segment the dataset based on specific columns when determining how to handle the missing values?
- Are outliers a factor in this dataset?

Remember preserving as much of the data as possible is the goal. That means dropping rows with a missing value somewhere might not be the best idea.

[Add you answer to this markdown cell]

##### I'm noticing that there are several columns that can't be effectively cleaned. One column specifically is Q4. The question is there to provide additional context about a job title, while the previous question (Q3) asks what the job title is, and is a required field. Only about a quarter of those surveyed responded to that question and by itself so the other 75% will show null values for that field. The timestamp field is one that would be hard to effectively clean and even if we could, it wouldn't provide much useful data. Q8 and Q9 are made up of almost entirely null values so I don't believe we're going to get much value out of either of those columns. There are some outliers in this dataset when it comes to salary. I think the primary issue there is that there is only one column for salary regardless of the currency.

##### To break down the changes I'm going to make, I'm listing out each change I'm going to make:
##### 1. Drop Q8, that column has 27,977 null values out of 28,108 total rows.
##### 2. Drop Q9, that column has 25,061 null values out of 28,108 total rows.
##### 3. Drop timestamp, very little if any value with this column, plus extremely difficult to clean.
##### 4. Drop Q4, we can work without the additional details about job titles. Also, there is very little data in that column.
##### 5. Drop rows that contain a currency other than USD. USD cover the large majority of the dataset (23,410 of 28,108) and the only alternative I can think of is to engineer some features to convert the other currencies to USD using the conversion rate for each currency, but that is outside of my skillset and what I was able to find on the internet.
##### 6. Remove commas from q5 and convert to float.

#### Implementation

Based on the plan the you described above, go ahead and clean up the dataset.

[Add as many code cell below here as needs]

In [33]:
df = df.drop(['q8', 'q9', 'timestamp', 'q4'], axis=1)


df.head()

q1                             q2  \
0  25-34   Education (Higher Education)   
1  25-34              Computing or Tech   
2  25-34  Accounting, Banking & Finance   
3  25-34                     Nonprofits   
4  25-34  Accounting, Banking & Finance   

                                         q3      q5      q6   q7  \
0        Research and Instruction Librarian  55,000     0.0  USD   
1  Change & Internal Communications Manager  54,600  4000.0  GBP   
2                      Marketing Specialist  34,000     NaN  USD   
3                           Program Manager  62,000  3000.0  USD   
4                        Accounting Manager  60,000  7000.0  USD   

              q10             q11          q12           q13          q14  \
0   United States   Massachusetts       Boston     5-7 years    5-7 years   
1  United Kingdom             NaN    Cambridge  8 - 10 years    5-7 years   
2              US       Tennessee  Chattanooga   2 - 4 years  2 - 4 years   
3             USA       Wisconsin    Milwaukee  8 - 10 years    5-7 years   
4              US  South Carolina   Greenville  8 - 10 years    5-7 years   

               q15         q16    q17  
0  Master's degree       Woman  White  
1   College degree  Non-binary  White  
2   College degree       Woman  White  
3   College degree       Woman  White  
4   College degree       Woman  White

In [34]:
df = df[df['q7'] == 'USD']
df.head()

q1                             q2                                  q3  \
0  25-34   Education (Higher Education)  Research and Instruction Librarian   
2  25-34  Accounting, Banking & Finance                Marketing Specialist   
3  25-34                     Nonprofits                     Program Manager   
4  25-34  Accounting, Banking & Finance                  Accounting Manager   
5  25-34   Education (Higher Education)      Scholarly Publishing Librarian   

       q5      q6   q7            q10             q11          q12  \
0  55,000     0.0  USD  United States   Massachusetts       Boston   
2  34,000     NaN  USD             US       Tennessee  Chattanooga   
3  62,000  3000.0  USD            USA       Wisconsin    Milwaukee   
4  60,000  7000.0  USD             US  South Carolina   Greenville   
5  62,000     NaN  USD            USA   New Hampshire      Hanover   

            q13          q14              q15    q16    q17  
0     5-7 years    5-7 years  Master's degree  Woman  White  
2   2 - 4 years  2 - 4 years   College degree  Woman  White  
3  8 - 10 years    5-7 years   College degree  Woman  White  
4  8 - 10 years    5-7 years   College degree  Woman  White  
5  8 - 10 years  2 - 4 years  Master's degree    Man  White

In [39]:
df['q5'] = df['q5'].str.replace(',', '')
df.head()

q1                             q2                                  q3  \
0  25-34   Education (Higher Education)  Research and Instruction Librarian   
2  25-34  Accounting, Banking & Finance                Marketing Specialist   
3  25-34                     Nonprofits                     Program Manager   
4  25-34  Accounting, Banking & Finance                  Accounting Manager   
5  25-34   Education (Higher Education)      Scholarly Publishing Librarian   

      q5      q6   q7            q10             q11          q12  \
0  55000     0.0  USD  United States   Massachusetts       Boston   
2  34000     NaN  USD             US       Tennessee  Chattanooga   
3  62000  3000.0  USD            USA       Wisconsin    Milwaukee   
4  60000  7000.0  USD             US  South Carolina   Greenville   
5  62000     NaN  USD            USA   New Hampshire      Hanover   

            q13          q14              q15    q16    q17  
0     5-7 years    5-7 years  Master's degree  Woman  White  
2   2 - 4 years  2 - 4 years   College degree  Woman  White  
3  8 - 10 years    5-7 years   College degree  Woman  White  
4  8 - 10 years    5-7 years   College degree  Woman  White  
5  8 - 10 years  2 - 4 years  Master's degree    Man  White

In [40]:
df['q5'] = df['q5'].astype(float)

In [43]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 23410 entries, 0 to 28107
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   q1      23410 non-null  object 
 1   q2      23350 non-null  object 
 2   q3      23409 non-null  object 
 3   q5      23410 non-null  float64
 4   q6      17539 non-null  float64
 5   q7      23410 non-null  object 
 6   q10     23410 non-null  object 
 7   q11     23045 non-null  object 
 8   q12     23347 non-null  object 
 9   q13     23410 non-null  object 
 10  q14     23410 non-null  object 
 11  q15     23254 non-null  object 
 12  q16     23268 non-null  object 
 13  q17     23260 non-null  object 
dtypes: float64(2), object(12)
memory usage: 2.7+ MB


#### Reflection
Write a short reflection (400-500 words) answering the following: 
- What were the biggest issues you encountered in the messy dataset?
- How did cleaning the dataset improve its usability for machine learning?
- What would happen if we trained a model on the messy dataset vs. the cleaned one?
- Do you feel you skewed or biased the dataset while cleaning it?

[Add you answer to this markdown cell]

 The biggest issues I encountered with this dataset had to do with the survey itself. The fact that they had free form questions created a challenge because those fields were designed to add context to a previous question or replace the previous question. That created features that had a high percentage of null values, making the input difficult to use. Another issue is that they asked for a salary and a currency, but did not separate the salary question for other currencies. They also used a few number ranges, which makes it quite hard to do any kind of analysis on. I couldn’t think of a way to clean those columns in a way that would be accurate. I could have changed them to the first number in the range, I saw a few people doing that online, but I didn’t feel that method would provide an accurate result. I guess it’s likely with 28,000 rows that they would balance out, but it still doesn’t feel quite right.
I think the cleaning I did will certainly make the data more usable. Now we have very few null values remaining. To get rid of the rest, I would engineer a feature to combine q5 and q6 to get a total overall salary and drop those columns. Converting the salary to float should allow us to run some analysis and build a model. I think it’s better than it was but to clean it up, I ended up losing some data, which could result in some inaccuracies.
I don’t think you really could train a model with the messy data. If you did try, I would guess you would get some strange results. I think there was an entry from Tanzania where the person made 2,000,000 per year in their local currency, but without anything separating that value from US dollars in the salary column it would create a massive outlier.
I think that the cleaning I did removed a lot of the skewness and bias from the dataset. Like I mentioned above, some of those other currencies were going to create some significant biases. There was a person on there that was paid in Rupees that said they made 10 Rupees per year. Removing everyone that isn’t being paid in USD surely created some bias.
Overall, I think it would be very hard to build a good working machine learning model with this data. I’m sure it’s exhausting as much as I’ve talked about the survey design but I’m positive they did not have data scientists in mind when they created it.


## Deliverables
Upload your Jupyter Notebook to your GitHub repo and then provide a link to that repo in Worlclass. 